File: december-2016/20161221-first_unique_character_in_a_string.rb
# Given a string, find the first non-repeating character in it and return its index. If it doesn't, return -1.
# s = "leetcode"
# return 0.
#
# s = "loveleetcode"
# return 2.

# pseudo code
# iterate through the string - O(n) time O(n) - space
# store character count in Hash
# return the first key with count == 1
# if no count == 1 return -1

def first_uniq_char(string)
  string_array = string.chars
  # make a hash with default value of 0
  char_hash = Hash.new(0)
  string_array.each do |char|
    char_hash[char] += 1
  end

  char_hash.delete_if{ |char, ct| ct > 1 }
  if char_hash.empty?
    p -1
  else
    p char_hash.first[0]
  end

end

# vroom vroom
first_uniq_char("loveleetcode")
# should return "v"
first_uniq_char("bobo")
# should return -1


# Big O Analysis
# Time - at least O(n) because we have to iterate through the string array once. Returning values from hash is just O(1)
# Space - O(n) because we are storing every character in a hash

File: december-2016/20161221-first_unique_character_in_a_string.py
# Source: Leetcode

# https://leetcode.com/problems/first-unique-character-in-a-string/

# ----

# Given a string, find the first non-repeating character in it and return its index. If it doesn't, return -1.

# **Examples:**


# s = "leetcode"
# return 0.
#
# s = "loveleetcode"
# return 2.


# **Note:** You may assume the string contains only lowercase letters.

# loop through the string and count (returns how many occurrences of 's' exist)

class Solution(object):
    def firstUniqChar(self, s):
      for x in s:
        if s.count(x) == 1:
          return [s.index(x)][0]
      else:
        return -1

# should return 0
test_one = "leetcode"
uniq_char = Solution()
print(uniq_char.firstUniqChar(test_one))

# should return 2
test_two = "loveleetcode"
print(uniq_char.firstUniqChar(test_two))

# should return -1
test_three = "uugghh"
print(uniq_char.firstUniqChar(test_three))

# should return -1
test_four = ""
print(uniq_char.firstUniqChar(test_four))

# big o
# time complexity
    # answer: O(n)
# space complexity
    # answer: O(n)
